src/engines/qwant.rs

```diff
@@ -0,0 +1,177 @@
+}
+
+#[derive(Deserialize, Debug)]
+struct QwantResult {
+    // NOTE: This object also contains `denied`, `total`, `items`, `filters`, `lastPage`,
+    //       `instrumentation`, `onlyProductAds`, and `topClassification` attributes,
+    //       which we currently don't care about.
+    /// Entries that should be shown to the user
+    items: QwantItems,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "snake_case")]
+#[serde(tag = "status", content = "data")]
+enum QwantApiResponse {
+    /// Success response
+    Success {
+        // NOTE: This object also contains `query` and `cache` attributes,
+        //       which we currently don't care about.
+        /// Actual results the search produced
+        result: QwantResult,
+    },
+    // TODO: Use the reported error messages
+    #[allow(unused)]
+    /// Error response
+    Error {
+        /// Machine-readable error code
+        error_code: i32,
+        #[serde(default)]
+        /// List of human-readable error messages
+        message: Vec<String>,
+    },
+}
+
+impl From<QwantApiResponse> for Result<QwantResult, EngineError> {
+    fn from(value: QwantApiResponse) -> Self {
+        match value {
+            QwantApiResponse::Success { result } => Ok(result),
+            QwantApiResponse::Error { .. } => Err(Report::new(EngineError::RequestError)),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SearchEngine for Qwant {
+    async fn results(
+        &self,
+        query: &str,
+        page: u32,
+        user_agent: &str,
+        client: &Client,
+        safe_search: u8,
+    ) -> Result<Vec<(String, SearchResult)>, EngineError> {
+        let results_per_page = 10;
+        let start_result = results_per_page * page;
+
+        let url = Url::parse_with_params(
+            "https://api.qwant.com/v3/search/web",
+            &[
+                ("q", Cow::from(query)),
+                ("count", results_per_page.to_string().into()),
+                ("locale", "en_US".into()),
+                ("offset", start_result.to_string().into()),
+                ("safesearch", safe_search.to_string().into()),
+                ("device", "desktop".into()),
+                ("tgb", "2".into()),
+                ("displayed", "true".into()),
+            ],
+        )
+        .change_context(EngineError::UnexpectedError)?;
+
+        let header_map = HeaderMap::try_from(&HashMap::from([
+            ("User-Agent".to_string(), user_agent.to_string()),
+            ("Referer".to_string(), "https://qwant.com/".to_string()),
+            ("Origin".to_string(), "https://qwant.com".to_string()),
+        ]))
+        .change_context(EngineError::UnexpectedError)?;
+
+        let result: QwantApiResponse = client
+            .get(url)
+            .headers(header_map)
+            .send()
+            .await
+            .change_context(EngineError::RequestError)?
+            .json()
+            .await
+            .change_context(EngineError::RequestError)?;
```

Comment:
I think you could have used the already available `json_as_bytes_from_upstream` function to fetch data and deserialize it using something like `bincode`. What do you think? 🙂 .